File: June-LeetCoding-Challenge/21-Dungeon-Game/Dungeon-Game.py
class Solution:
    
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        a = len(dungeon)
        b = len(dungeon[0])
        dp = [[0 for x in range(b)] for y in range(a)]
        if dungeon[-1][-1]<=0:
            dp[-1][-1] = 1 - dungeon[-1][-1]
        else:
            dp[-1][-1] = 1
        for i in range(b-2,-1,-1):
            dp[-1][i] = max(1,dp[-1][i+1]-dungeon[-1][i])
        for j in range(a-2,-1,-1):
            dp[j][-1] = max(1,dp[j+1][-1]-dungeon[j][-1])
        for i in range(a-2,-1,-1):
            for j in range(b-2,-1,-1):
                dp[i][j] = max(1,min(dp[i+1][j],dp[i][j+1])-dungeon[i][j])
        return dp[0][0]

File: June-LeetCoding-Challenge/21-Dungeon-Game/Dungeon-Game.cpp
class Solution {
public:
    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        int n = dungeon.size();
        int m = dungeon[0].size();
        vector<vector<int>> dp(n+1, vector<int>(m+1, 1000000000));
        dp[n-1][m] = dp[n][m-1] = 1;
        
        for (int i = n-1; i>=0; i--)
            for (int j = m-1; j>=0; j--)
                dp[i][j] = max(min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j], 1);
        
        return dp[0][0];
    }
};